File: HealthLink_BE/src/main/java/com/HealthLink/entity/PharmacyConsultationRequest.java
package com.HealthLink.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Nationalized;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "PharmacyConsultationRequests")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PharmacyConsultationRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "RequestID")
    private Integer requestId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "PatientID", nullable = false)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private Patient patient;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "PharmacyID", nullable = false)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private Pharmacy pharmacy;

    @OneToOne(mappedBy = "consultationRequest", fetch = FetchType.LAZY)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private PharmacyOrder order;

    @OneToMany(mappedBy = "consultationRequest", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("createdAt ASC, requestPrescriptionId ASC")
    @Builder.Default
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private List<PharmacyConsultationRequestPrescription> requestPrescriptions = new ArrayList<>();

    @Column(length = 2000)
    private String symptoms;

    @Column(length = 4000)
    private String description;

    @Column(length = 1000)
    private String allergies;

    @Column(columnDefinition = "TEXT")
    private String attachments;

    @Column(length = 2000)
    private String additionalNotes;

    @Column(length = 50)
    private String preferredDeliveryType;

    @Column(length = 50)
    private String deliveryType;

    @Nationalized
    @Column(length = 500)
    private String deliveryAddress;

    private Double deliveryLatitude;

    private Double deliveryLongitude;

    @Column(length = 30)
    private String deliveryPhoneNumber;

    @Column(length = 50)
    private String deliveryAddressSource;

    @Column(length = 50)
    private String requestType;

    @Column(length = 50, nullable = false)
    private String status;

    @Column(length = 36)
    private String chatRoomId;

    @Column(length = 2000)
    private String pharmacyNotes;

    @Column(length = 2000)
    private String patientFollowUpNotes;

    @Column(name = "CreatedAt", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "UpdatedAt", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        if (createdAt == null) {
            createdAt = now;
        }
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: HealthLink_MB/lib/models/pharmacy/pharmacy_consultation_request.dart
class PharmacyConsultationRequest {
  final int requestId;
  final String patientId;
  final String patientName;
  final String? pharmacyId;
  final String? pharmacyName;
  final String? pharmacyUserId;
  final String? symptoms;
  final String? description;
  final String? allergies;
  final List<String>? attachments;
  final String? additionalNotes;
  final String? preferredDeliveryType;
  final String? requestType;
  final String? deliveryType;
  final String? deliveryAddress;
  final String? deliveryPhoneNumber;
  final String status;
  final String? chatRoomId;
  final String? pharmacyNotes;
  final String? patientFollowUpNotes;
  final List<int>? prescriptionHeaderIds;
  final int? pharmacyOrderId;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const PharmacyConsultationRequest({
    required this.requestId,
    required this.patientId,
    required this.patientName,
    this.pharmacyId,
    this.pharmacyName,
    this.pharmacyUserId,
    this.symptoms,
    this.description,
    this.allergies,
    this.attachments,
    this.additionalNotes,
    this.preferredDeliveryType,
    this.requestType,
    this.deliveryType,
    this.deliveryAddress,
    this.deliveryPhoneNumber,
    required this.status,
    this.chatRoomId,
    this.pharmacyNotes,
    this.patientFollowUpNotes,
    this.prescriptionHeaderIds,
    this.pharmacyOrderId,
    required this.createdAt,
    this.updatedAt,
  });

  factory PharmacyConsultationRequest.fromJson(Map<String, dynamic> json) {
    return PharmacyConsultationRequest(
      requestId: json['requestId'] as int? ?? 0,
      patientId: json['patientId'] as String? ?? '',
      patientName: json['patientName'] as String? ?? '',
      pharmacyId: json['pharmacyId'] as String?,
      pharmacyName: json['pharmacyName'] as String?,
      pharmacyUserId: json['pharmacyUserId'] as String?,
      symptoms: json['symptoms'] as String?,
      description: json['description'] as String?,
      allergies: json['allergies'] as String?,
      attachments: (json['attachments'] as List<dynamic>?)
          ?.map((e) => e.toString())
          .toList(),
      additionalNotes: json['additionalNotes'] as String?,
      preferredDeliveryType: json['preferredDeliveryType'] as String?,
      requestType: json['requestType'] as String?,
      deliveryType: json['deliveryType'] as String?,
      deliveryAddress: json['deliveryAddress'] as String?,
      deliveryPhoneNumber: json['deliveryPhoneNumber'] as String?,
      status: json['status'] as String? ?? 'PENDING',
      chatRoomId: json['chatRoomId'] as String?,
      pharmacyNotes: json['pharmacyNotes'] as String?,
      patientFollowUpNotes: json['patientFollowUpNotes'] as String?,
      prescriptionHeaderIds: (json['prescriptionHeaderIds'] as List<dynamic>?)
          ?.map((e) => e as int)
          .toList(),
      pharmacyOrderId: json['pharmacyOrderId'] as int?,
      createdAt: _parseDateTime(json['createdAt']) ?? DateTime.now(),
      updatedAt: _parseDateTime(json['updatedAt']),
    );
  }

  static DateTime? _parseDateTime(dynamic value) {
    if (value == null) return null;
    if (value is DateTime) return value;
    if (value is String) {
      try {
        return DateTime.parse(value);
      } catch (_) {
        return null;
      }
    }
    return null;
  }
}
